Add: Joined Colony List

willm30 opened the pull request:
Add a popover on the Active Colony avatar showing a list of subscribed colonies.

## Description

This PR focuses on Screen 2 of the [Figma Design](https://www.figma.com/file/DjVCPq7LpkjEMEfQZtKgBdC1/Style-Guide?node-id=0%3A1), namely the list of subscribed colonies on mobile. 

**New stuff** ✨

* `SubscribedColoniesDropdown`

**Changes** 🏗

* Passing the `path` to `SusbcribedColoniesList` so that it knows which of the colonies is the active one, and thus which to render as the menu icon. 


## Screenshots

![subscribed-cols](https://user-images.githubusercontent.com/64402732/175099115-adaea37a-e6b5-4602-8e80-55ac62d191c5.png)



Resolves #3520

willm30 commented on src/modules/dashboard/components/SubscribedColoniesList/SubscribedColoniesDropdown.tsx:
```diff
@@ -0,0 +1,93 @@
+        ],
+      }}
+    >
+      <NavLink
+        activeClassName={styles.activeColony}
+        className={styles.itemLink}
+        title={activeColony.colonyName}
+        to={`/colony/${activeColony.colonyName}`}
+      >
+        <div className={styles.itemImage}>
+          <ColonyAvatar
+            colony={activeColony}
+            colonyAddress={activeColony.colonyAddress}
+            size="xs"
+          />
+        </div>
+      </NavLink>
```
I wanted to extract this into a `SubscribedColony` component, as it's also used in the desktop render of the `SubscribedColoniesList`. However, extracting it into its own component and passing that component as the child of Popover broke the popover (dropdown menu didn't open on click). Would need to spend an amount of time disporportionate to the benefit gained to try to figure out why (tried a little bit but to no avail), but note here in case its actually an easy fix.

rdig replied:
Try wrapping the `Popover` in a `div` before rendering it

I _think_ that might solve your issue

willm30 replied:
I played around with it a little bit, putting `div`s inside the component, around the component, passing through the ref provided by the `Popover` element, using forward ref... It seems to be specifically due to this `Nav Link` block because rending anything else inside the `Popover` works. Puzzling.

willm30 commented on src/modules/dashboard/components/SubscribedColoniesList/SubscribedColoniesList.tsx:
```diff
@@ -42,7 +52,16 @@ const SubscribedColoniesList = () => {
+            {activeColony && (
+              <SubscribedColoniesDropdown
+                activeColony={activeColony}
+                coloniesList={data!.user.processedColonies}
```
If `data` is undefined, `activeColony` will be undefined.

rdig commented on src/modules/dashboard/components/SubscribedColoniesList/SubscribedColoniesDropdown.tsx:
```diff
@@ -0,0 +1,93 @@
+                    title={colony.colonyName}
+                    to={`/colony/${colony.colonyName}`}
+                  >
+                    <div className={styles.dropdownItem}>
+                      <div className={styles.itemImage}>
+                        <ColonyAvatar
+                          colony={colony}
+                          colonyAddress={colony.colonyAddress}
+                          size="xs"
+                        />
+                      </div>
```
![Screenshot from 2022-06-23 22-30-15](https://user-images.githubusercontent.com/1193222/175382732-7119fb83-59aa-4c53-8ee7-542e5c75a5dd.png)

I think we need less padding between the elements. This is more of a think for Product, rather than your implementation

rdig commented on src/modules/dashboard/components/SubscribedColoniesList/SubscribedColoniesDropdown.tsx:
```diff
@@ -0,0 +1,93 @@
+    <Popover
+      content={() => {
+        return (
+          <DropdownMenu>
```
If you have a really long number of colonies, it will break the view / dropdown.

See example below:

![mobile-view-more-colonies-break-view](https://user-images.githubusercontent.com/1193222/175383202-5e077ca3-54ab-4556-982f-94d13d88accf.gif)

The dropdown should have a fixed height _(max-height)_ after which it should just allow scrolling between the items

rdig replied:
Less colonies work fine though

![mobile-view-less-colonies-ok](https://user-images.githubusercontent.com/1193222/175383588-3eb67b22-7a1f-44ab-9df4-740c2eba3858.gif)

rdig reviewed (changes requested):
Nice work Will!

I'm mostly good with this, however you have to deal with the dropdown, if the number of colonies is higher.

rdig reviewed (changes requested):
Mostly good, but two things need to be addressed.

1. The colonies menu in the mobile view needs to be aligned to the bottom of the active colony, while your current implementation aligns it to the right of it.

Current:
![Screenshot from 2022-06-24 16-44-22](https://user-images.githubusercontent.com/1193222/175548960-07cddd8a-915a-414f-bb5a-130f36a8b46a.png)
Specs:
![Screenshot from 2022-06-24 16-44-38](https://user-images.githubusercontent.com/1193222/175548968-80859ae3-2285-4b6a-b897-7e76d2fd509c.png)

2. If you are not subscribed to the current active colony, you cannot access the colonies menu in the mobile view. 

See demo:
![mobile-view-bug-not-subscribed-active-colony](https://user-images.githubusercontent.com/1193222/175549274-3e345a85-a766-4f4f-be7d-f478dbd58add.gif)

willm30 commented:
@rdig I've explicitly made the placement of the popover `bottom` and fixed the issue with the disappearing list (see below). 
I also cleaned up the logic regarding the active colony filter because using the string method `includes` on the path string was occasionally returning more than one result. 

![subscribed-cols](https://user-images.githubusercontent.com/64402732/175774348-7a232ec8-3d8c-40c4-bb71-5e0ae0cf9772.gif)

arrenv commented on src/modules/dashboard/components/SubscribedColoniesList/SubscribedColoniesList.css:
```diff
@@ -75,12 +80,48 @@
+
+  .scrollableContainer li[class*="DropdownMenuItem"] > *,
+  .scrollableContainer li[class*="DropdownMenuItem"]:only-child > * {
+    padding: 13px 0px 13px 11px;
```
Can we make this `padding: 6px 0 6px 12px;`

arrenv replied:
We can also add a class to the `main` of this Dropdown with `padding: 6px 0;`

arrenv commented on src/modules/dashboard/components/SubscribedColoniesList/SubscribedColoniesList.css:
```diff
@@ -75,12 +80,48 @@
+
+  .scrollableContainer li[class*="DropdownMenuItem"] figure {
+    border-radius: 50%;
+    box-shadow: 0px 1px 6px rgba(62, 118, 244, 0.14);
```
This color can be replaced with `var(--drop-shadow)`

arrenv reviewed (changes requested):
Nice one @willm30 a couple of comments.

willm30 commented:
@arrenv Padding added and kebab menu enlarged!

rdig commented on src/modules/pages/components/RouteLayouts/Default/Default.tsx:
```diff
@@ -40,7 +40,7 @@ const Default = ({
     hasSubscribedColonies ? (
       <div className={styles.coloniesList}>
-        <SubscribedColoniesList />
+        <SubscribedColoniesList path={location.pathname} />
```
This has a memory leak somewhere, and now I'm not 100% sure, but I think this PR introduced it.

![Screenshot from 2022-06-29 14-58-51](https://user-images.githubusercontent.com/1193222/176431374-b0c3f55f-4e89-4edf-9955-5ed8bf500e06.png)

willm30 replied:
@rdig I have fixed this in this PR https://github.com/JoinColony/colonyDapp/pull/3568

willm30 replied:
@rdig It wasn't this PR so created a separate branch to merge straight into the main feature branch

rdig reviewed (approved):
Looks all good now Will. Nice work! :100: 

![Screenshot from 2022-06-29 14-59-42](https://user-images.githubusercontent.com/1193222/176431399-11ab9966-dfc5-4558-8c72-48ae699b9344.png)
